rfcs/rfc-replace-SQLContinuum-with-SQLAlchemy-history.md

```diff
@@ -0,0 +1,105 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- Target Major Version: (EPIC or User Story TAG|Link)
+- Reference Issues: (fill in existing related issues, if any)
+- Entity Issue: (leave this empty)
+- Implementation PR: (leave this empty)
+
+# Summary
+
+SQLContinuum's main developer has stopped maintianing the package, and although there are numerous committers, the latest breaking changes to Flask and SQLAlchemy have effectively halted development on it.
+BCRegistry does not use many features of SQLContinuum, other than to create a _version_ table, and use the SQLContinuum Transaction table. The SQLContinuum version methods are not used.
+
+# Basic example
+
+models/db.py to wrap versioning into the Flask-SQLAlchemy managed session
+```python
+from flask_sqlalchemy import SQLAlchemy
+from .versioned_history import versioned_session
+db = SQLAlchemy()  # pylint: disable=invalid-name
+versioned_session(db.session)
+```
+
+models/submission.py  combining `Versioned` and `db.Model` to make a versioned class
+```python
+from .db import db
+from .versioned_history import Versioned
+
+class Submission(Versioned, db.Model):
+    """Stores a submission of JSON data."""
+    __tablename__ = "submission"
+    id = db.Column(db.Integer, primary_key=True)
+```
+
+This adds a `version` column to the _base_ table, and creates a versioned table using `__tablename__`\_history which is a copy of the _base_ table and adds a _changed_ column as well.
+
+There are no special helper methods to access the history.
+
+
+# Motivation
+
+To allow services using SQLContinuum to migrate to the newer versions of Python, Flask and SQLAlchemy.
+
+# Detailed design
+
+### Entities Specific
+SQLContinuum uses a Transaction table, and stamps the txn_id into every versioned record altered in the transactions. To accomodate that design the txn_id was added to the _Filings_ table.
+An equivalent would be to add the filing_id to every _base_ table and stamp any changes made by that filing.
+
+```python
+from .db import db
+from .versioned_history import Versioned
+
+class Filing(Versioned, db.Model):
+    """Stores a submission of JSON data."""
+    __tablename__ = 'filings'
+    id = db.Column(db.Integer, primary_key=True)
+```
+
+```python
+from .db import db
+from .versioned_history import Versioned
+
+class Business(Versioned, db.Model):
+    """Stores a submission of JSON data."""
+    __tablename__ = 'businesses'
+    id = db.Column(db.Integer, primary_key=True)
+    identifier = db.Column('identifier',
+                           db.String(10),
+                           index=True)
+    founding_date = db.Column('founding_date',
+                              db.DateTime(timezone=True),
+                              default=datetime.utcnow)
+
+    # filing_id that made the latest change
+    change_filing_id = db.Column('change_filing_id',
+                                 db.Integer,
+                                 db.ForeignKey('filings.id'))
```

Comment:
There are transactions currently that have no matching `filing_id` in the existing implementation.  What would we do if we go with this approach of using the `change_filing_id`?

Here's an example I found in Prod.  I think the transaction was for updating the `businesses.tax_id` field.

![image](https://github.com/bcgov/entity/assets/6685223/4d2085af-c90a-4b4b-b28e-b976071af5ac)

Reply:
That was probably a manual update, that did not use the Queues (which is the recommended approach). All changes to data are attached to a filing to be displayed in the ledger.

Reply:
So I just talked to Vysakh and entity-bn does not link back to a filing id when updating the `businesses.tax_id` for example.

Reply:
-->> https://github.com/bcgov/lear/blob/main/legal-api/src/legal_api/resources/v2/internal_services.py#L64

Reply:
As you can see, it creates a RegistrarsNotation, which will appear on the ldeger.

Reply:
ok, that makes sense.  I still see the `businesses.tax_id` being updated with no RegistrarsNotation [here](https://github.com/bcgov/lear/blob/main/queue_services/entity-bn/src/entity_bn/bn_processors/registration.py#L207-L223) though.  Should we added a RegistrarsNotation entry here and tie to the transaction id?  

And since we do have instances in Prod already where we will have no value for the `change_filing_id`, what do we want to do?  I think for the tax_id example I gave, we could backfill the db with registrar notation entries.  But in other instances, is it ok to just have a version entry with a null `change_filing_id`?

Reply:
possibly.  we could also tie it to the filing that created the BN.
I know ops was just banging in some of the bn-moves, but they should now be hitting the end-point Vysakh added which will "just do the right thing".

Reply:
while migrating it should tie to registration filing